### DISCUSSION TITLE
Why length of top_class_scores != top_k?

### DISCUSSION THREAD

**Discussion by aswincandra** [aurelio-labs/semantic-router#325] [Q&A]
In [RouteLayer](https://github.com/aurelio-labs/semantic-router/blob/main/semantic_router/layer.py) class, there is `_retrieve_top_route` function that returns `top_class_scores`. Why length of top_class_scores is not same with the top_k? So Semantic Router doesn't actually output exactly top-k utterances?

**Comment by octocat20** [comment]
top_k and top_class_scores measure different things.

top_k is the number of nearest utterances fetched from the index. Those hits are then grouped by route. top_class_scores is only the scores for utterances that belong to the winning route, not all top_k neighbors.

So if top_k=5 and three neighbors are route A and two are route B, the winning route might only have 3 scores. That is expected. Semantic Router ranks routes after aggregation (mean by default), not "return exactly top-k utterances".
